### PR TITLE
Fix compiler warnings, fix bugs with non-total pattern matches

### LINKED ISSUE #11: Executing bracket notation with dot runs into non-exhaustive pattern match
Path:
```
$['two.some']
```

Input:
```
{"one": {"key": "value"}, "two": {"some": "more", "key": "other value"}, "two.some": "42"}
```
Expected output:
```
"42"
```
Error:
```
json-path-comparison: src/Data/JSONPath/Execute.hs:(29,1)-(63,29): Non-exhaustive patterns in function executeJSONPathElement

```

Split from #6

### LINKED ISSUE #12: Bracket notation with quoted array slice literal fails to parse
Path:
```
$[':']
```
Input:
```
{":": "value", "another": "entry"}
```
Expected output:
```
"value"
```
Error:
```
json-path-comparison: searchBeginningWithSlice: string
CallStack (from HasCallStack):
  error, called at app/Main.hs:36:7 in main:Main
Invalid JSONPath: $[':']
 Error: searchBeginningWithSlice: string
```

Split from #6

### LINKED ISSUE #13: Bracket notation with quoted current object literal fails to parse
Path:
```
$['@']
```
  Input:
  ```
  {"@": "value", "another": "entry"}
  ```
  Expected output:
  ```
  "value"
  ```
  Error:
  ```
  json-path-comparison: searchBeginningWithSlice: string
  CallStack (from HasCallStack):
    error, called at app/Main.hs:36:7 in main:Main
  Invalid JSONPath: $['@']
   Error: searchBeginningWithSlice: string
  ```

Split from #6

### PR DESCRIPTION
**Pull request by schrammc-ebot7**
This commit fixes a variety of issues:

- The parser for field access with ['fieldname'] notation was too strict and did not allow cases where there were spaces betewen the brackets and single-quotes
- The parser did not allow access to fieldnames with special chars like .,#, etc.
- There were many compiler warnings when run with -Wall
- There library failed due to non-total pattern match in executeCondition

Also this commit adds the following condition operators:
- `>`
- `<=`
- `<=`

This fixes #11 
This fixes #12
This fixes #13 
This will resolve much of #6

### CODE REVIEW

File: src/Data/JSONPath/Execute.hs [hunk 1]
```diff
@@ -64,13 +64,22 @@ executeJSONPathElement s@(Search js) val =
...
 valMap _ v = pure $ ResultError $ "Expected object or array, found " <> (encodeJSONToString v)
 
 executeCondition :: Value -> Condition -> Literal -> Bool
-executeCondition (Number n1) Equal (LitNumber n2) = n1 == (fromInteger $ toInteger n2)
+executeCondition val NotEqual lit = not $ executeCondition val Equal lit
+executeCondition (Number n1) Equal (LitNumber n2) = n1 == realToFrac n2
 executeCondition (String s1) Equal (LitString s2) = s1 == s2
+executeCondition _ Equal _ = False
+executeCondition val GreaterThan lit =
+  not (executeCondition val SmallerThan lit) &&  not (executeCondition val Equal lit)
```
Comment (schrammc-ebot7) [inline src/Data/JSONPath/Execute.hs:77]: :information_source: The idea here is to implement all operators in terms of `==` and `<`

File: src/Data/JSONPath/Types.hs [hunk 2]
```diff
@@ -19,9 +19,11 @@ data Condition = Equal
                | NotEqual
                | GreaterThan
                | SmallerThan
+               | GreaterEqualThan
+               | SmallerEqualThan
   deriving (Show, Eq)
 
-data Literal = LitNumber Int
+data Literal = LitNumber Double
```
Comment (schrammc-ebot7) [inline src/Data/JSONPath/Types.hs:26]: :information_source:  I think being able to compare doubles here makes sense

File: src/Data/JSONPath/Types.hs [hunk 3]
```diff
@@ -36,7 +38,6 @@ data SliceElement = SingleIndex Int
   deriving (Show, Eq)
 
 data JSONPathElement  = KeyChild Text
-                      | KeyChildren [Text]
```
Comment (schrammc-ebot7) [inline src/Data/JSONPath/Types.hs:39]: :information_source: This was the cause for `$['two.some']` not working.

File: test/resources/json-path-tests/FilterExpressionSubscriptOperator.json [hunk 4]
```diff
@@ -1,32 +1,97 @@
...
+                       "path" : "$[?(@.textval <= \"b\")]",
+                       "result" : [{ "textval" : "a" }, { "textval" : "b" }]
+                   },
+                   {
+                       "path" : "$[?(@.val == \"1\")]",
+                       "result" : []
+                   }
+
+        ]
+
```
[the remark this answers is not in the document]
Comment (schrammc-ebot7) [reply test/resources/json-path-tests/FilterExpressionSubscriptOperator.json:96]: I think some of the other JSON-files actually used `\t` to indent. Should I switch them all to 4space?

[the remark this answers is not in the document]
Comment (akshaymankar) [reply test/resources/json-path-tests/FilterExpressionSubscriptOperator.json]: Thanks for pointing it out, I dealt with it.

**Review by akshaymankar — APPROVED** [review]

**Comment by akshaymankar** [comment]
Thanks again @schrammc-ebot7 for the PR!

**Comment by schrammc-ebot7** [comment]
Thanks for fixing the last issues with the PR, didn't get around to it yesterday :+1:

**Pull request merged**